File: src/main/java/com/github/tartaricacid/netmusic/api/NetWorker.java
package com.github.tartaricacid.netmusic.api;

import com.github.tartaricacid.netmusic.NetMusic;
import com.github.tartaricacid.netmusic.config.GeneralConfig;
import org.apache.commons.lang3.StringUtils;

import javax.annotation.Nullable;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.*;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Collection;
import java.util.Map;

/**
 * @author 内个球
 */
public class NetWorker {

    public static String get(String url, Map<String, String> requestPropertyData) throws IOException {
        StringBuilder result = new StringBuilder();
        URL urlConnect;

        try {
            urlConnect = new URL(url);
            URLConnection connection = urlConnect.openConnection(getProxyFromConfig());

            Collection<String> keys = requestPropertyData.keySet();
            for (String key : keys) {
                String val = requestPropertyData.get(key);
                connection.setRequestProperty(key, val);
            }

            connection.setConnectTimeout(12000);
            connection.setDoInput(true);

            try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(connection.getInputStream(),
                    StandardCharsets.UTF_8))
            ) {
                String line;
                while ((line = bufferedReader.readLine()) != null) {
                    result.append(line);
                }
            }
        } catch (IOException e) {
            NetMusic.LOGGER.error(e);
            throw e;
        } finally {

        }

        return result.toString();
    }

    //    public static String get(String url, Map<String, String> requestPropertyData) throws IOException {
    //        StringBuilder result = new StringBuilder();
    //        HttpURLConnection connection = null;
    //
    //        try {
    //            URL urlConnect = new URL(url);
    //            connection = (HttpURLConnection) urlConnect.openConnection(getProxyFromConfig());
    //            connection.setRequestMethod("GET");
    //            connection.setConnectTimeout(12000);
    //            connection.setReadTimeout(12000);
    //            connection.setDoInput(true);
    //
    //            // Set request properties
    //            for (Map.Entry<String, String> entry : requestPropertyData.entrySet()) {
    //                connection.setRequestProperty(entry.getKey(), entry.getValue());
    //            }
    //
    //            // Handle the response
    //            int responseCode = connection.getResponseCode();
    //            if (responseCode >= 200 && responseCode < 300) {
    //                Charset charset = StandardCharsets.UTF_8; // Default
    //                String contentType = connection.getContentType();
    //                if (contentType != null) {
    //                    for (String param : contentType.replace(" ", "").split(";")) {
    //                        if (param.startsWith("charset=")) {
    //                            charset = Charset.forName(param.split("=", 2)[1]);
    //                            break;
    //                        }
    //                    }
    //                }
    //
    //                try (BufferedReader bufferedReader = new BufferedReader(
    //                        new InputStreamReader(connection.getInputStream(), charset))) {
    //                    String line;
    //                    while ((line = bufferedReader.readLine()) != null) {
    //                        result.append(line);
    //                    }
    //                }
    //            } else {
    //                // Handle error responses
    //                throw new IOException("HTTP error code: " + responseCode);
    //            }
    //
    //        } finally {
    //            if (connection != null) {
    //                connection.disconnect();
    //            }
    //        }
    //
    //        return result.toString();
    //    }

    @Nullable
    public static String getRedirectUrl(String url, Map<String, String> requestPropertyData) throws IOException {
        URL urlConnect = new URL(url);
        URLConnection connection = urlConnect.openConnection(getProxyFromConfig());
        Collection<String> keys = requestPropertyData.keySet();
        for (String key : keys) {
            String val = requestPropertyData.get(key);
            connection.setRequestProperty(key, val);
        }

        connection.setConnectTimeout(3_000);
        connection.setReadTimeout(5_000);
        return connection.getHeaderField("Location");
    }

    public static String post(String url, String param, Map<String, String> requestPropertyData) throws IOException {
        StringBuilder result = new StringBuilder();
        BufferedReader bufferedReader;
        PrintWriter printWriter;

        URL urlConnect = new URL(url);
        URLConnection connection = urlConnect.openConnection(getProxyFromConfig());

        Collection<String> keys = requestPropertyData.keySet();
        for (String key : keys) {
            String val = requestPropertyData.get(key);
            connection.setRequestProperty(key, val);
        }
        connection.setConnectTimeout(12000);
        connection.setDoOutput(true);
        connection.setDoInput(true);

        printWriter = new PrintWriter(connection.getOutputStream());
        printWriter.print(param);
        printWriter.flush();

        bufferedReader = new BufferedReader(new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
        String line;
        while ((line = bufferedReader.readLine()) != null) {
            result.append(line);
        }

        bufferedReader.close();
        printWriter.close();

        return result.toString();
    }

    private static Proxy getProxyFromConfig() {
        Proxy.Type proxyType = GeneralConfig.PROXY_TYPE.get();
        String proxyAddress = GeneralConfig.PROXY_ADDRESS.get();
        if (proxyType == Proxy.Type.DIRECT || StringUtils.isBlank(proxyAddress)) {
            return Proxy.NO_PROXY;
        }

        String[] split = proxyAddress.split(":", 2);
        if (split.length != 2) {
            return Proxy.NO_PROXY;
        }
        return new Proxy(proxyType, new InetSocketAddress(split[0], Integer.parseInt(split[1])));
    }
}

File: src/main/java/com/github/tartaricacid/netmusic/api/NetWorker.kt
package com.github.tartaricacid.netmusic.api

import com.github.tartaricacid.netmusic.NetMusic
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.apache.http.client.config.RequestConfig
import org.apache.http.client.methods.CloseableHttpResponse
import org.apache.http.client.methods.HttpGet
import org.apache.http.client.methods.HttpPost
import org.apache.http.entity.ContentType
import org.apache.http.entity.StringEntity
import org.apache.http.impl.client.DefaultHttpRequestRetryHandler
import org.apache.http.impl.client.HttpClients
import org.apache.http.impl.conn.PoolingHttpClientConnectionManager
import org.apache.http.util.EntityUtils
import java.io.*
import java.net.*
import java.nio.charset.StandardCharsets

/**
 * NetWorker is a singleton object responsible for handling HTTP GET and POST requests.
 * It leverages Apache HttpClient for executing requests and supports both synchronous
 * and asynchronous operations using Kotlin coroutines.
 *
 * @author 内个球
 */
object NetWorker {

    private val connectionManager = PoolingHttpClientConnectionManager().apply {
        maxTotal = 10
        defaultMaxPerRoute = 10
    }

    private val HTTP_CLIENT = HttpClients.custom()
        .setConnectionManager(connectionManager)
        .setRetryHandler(DefaultHttpRequestRetryHandler(3, true))
        .build()

    private val DEFAULT_REQUEST_CONFIG = RequestConfig.custom()
        .setConnectTimeout(10_000) // 10 seconds
        .setConnectionRequestTimeout(10_000)  // 10 seconds
        .build()

    /**
     * Executes an asynchronous HTTP GET request.
     *
     * @param url The URL to send the GET request to.
     * @param requestProperties A map of request headers to include in the GET request.
     * @return The response body as a [String].
     * @throws NetWorkerException If the GET request fails.
     */
    suspend fun getAsync(url: String, requestProperties: Map<String, String?>): String = withContext(Dispatchers.IO) {
        get(url, requestProperties)
    }

    /**
     * Executes an asynchronous HTTP POST request.
     *
     * @param url The URL to send the POST request to.
     * @param param The request body as a [String]. Can be null.
     * @param requestProperties A map of request headers to include in the POST request.
     * @return The response body as a [String].
     * @throws NetWorkerException If the POST request fails.
     */
    suspend fun postAsync(url: String, param: String?, requestProperties: Map<String, String?>): String =
        withContext(Dispatchers.IO) {
            post(url, param, requestProperties)
        }

    /**
     * Executes a synchronous HTTP GET request.
     *
     * @param url The URL to send the GET request to.
     * @param requestProperties A map of request headers to include in the GET request.
     * @return The response body as a [String].
     * @throws NetWorkerException If the GET request fails or returns a non-2xx status code.
     */
    @Throws(NetWorkerException::class)
    fun get(url: String, requestProperties: Map<String, String?>): String {
        // Create the HTTP GET request
        val httpGet = HttpGet(url).apply {
            // Set headers from the requestProperties map
            requestProperties.forEach { (key, value) ->
                value?.let { setHeader(key, it) }
            }

            config = DEFAULT_REQUEST_CONFIG
        }

        try {
            // Execute the request using the HTTP client
            HTTP_CLIENT.execute(httpGet).use { response ->
                val statusCode = response.statusLine.statusCode
                val entity = response.entity ?: throw NetWorkerException("No response entity for GET request to $url")
                val responseBody = EntityUtils.toString(entity, StandardCharsets.UTF_8)

                if (statusCode in 200..299) {
                    return responseBody
                }

                throw NetWorkerException("GET request to $url returned status $statusCode: $responseBody")
            }
        } catch (e: IOException) {
            // Log the error with URL and request properties
            NetMusic.LOGGER.error("Failed to GET from URL: $url with properties: $requestProperties", e)
            throw NetWorkerException("GET request failed for URL: $url", e)
        }
    }


    /**
     * Retrieves the redirect URL from an HTTP response if a redirect is present.
     *
     * @param url The URL to send the GET request to.
     * @param requestProperties A map of request headers to include in the GET request.
     * @return The redirect URL as a [String] if a redirect status is received; otherwise, null.
     * @throws NetWorkerException If the request fails.
     */
    @Throws(NetWorkerException::class)
    fun getRedirectUrl(url: String, requestProperties: Map<String, String?>): String? {
        // Configure the request to not follow redirects automatically
        val requestConfig = RequestConfig.custom()
            .setRedirectsEnabled(false)
            .setConnectTimeout(3000)
            .setSocketTimeout(5000)
            .build()

        val httpGet = HttpGet(url).apply {
            requestProperties.forEach { (key, value) ->
                value?.let { setHeader(key, it) }
            }

            config = requestConfig
        }

        HTTP_CLIENT.execute(httpGet).use { response ->
            val statusCode = response.statusLine.statusCode
            if (statusCode in 300..399) {
                return response.getFirstHeader("Location")?.value
            }

            return null
        }
    }

    /**
     * Executes a synchronous HTTP POST request.
     *
     * @param url The URL to send the POST request to.
     * @param param The request body as a [String]. Can be null.
     * @param requestProperties A map of request headers to include in the POST request.
     * @param contentType The content type of the request body. Defaults to [ContentType.APPLICATION_JSON].
     * @return The response body as a [String]. Returns an empty string if the response has no entity.
     * @throws NetWorkerException If the POST request fails or returns a non-2xx status code.
     */
    @Throws(NetWorkerException::class)
    fun post(
        url: String,
        param: String?,
        requestProperties: Map<String, String?>,
        contentType: ContentType = ContentType.APPLICATION_JSON
    ): String {
        val requestConfig = RequestConfig.custom()
            // 12 seconds
            .setConnectTimeout(12_000)
            .setSocketTimeout(12_000)
            .build()

        // Create an HttpPost request with the specified URL
        val httpPost = HttpPost(url).apply {
            this.config = requestConfig
            // Set headers from requestPropertyData map
            requestProperties.forEach { (key, value) ->
                this.setHeader(key, value)
            }

            // Set the POST request body
            val entity = StringEntity(param, contentType)
            this.entity = entity
        }

        try {
            // Execute the POST request
            HTTP_CLIENT.execute(httpPost).use { response: CloseableHttpResponse ->
                // Check for a successful response (status code 2xx)
                val statusCode = response.statusLine.statusCode

                if (statusCode in 200..299) {
                    // Get the response entity
                    val entity = response.entity ?: return ""
                    // Convert the entity content to a String
                    return EntityUtils.toString(entity, StandardCharsets.UTF_8)
                }

                // Handle non-successful status codes as needed
                throw NetWorkerException("Unexpected response status: $statusCode for POST request to $url")
            }
        } catch (e: IOException) {
            // Log the error with URL and request properties
            NetMusic.LOGGER.error("Failed to POST to URL: $url with properties: $requestProperties", e)
            // Rethrow the exception to be handled by the caller
            throw NetWorkerException("POST request failed for URL: $url", e)
        }
    }

}